yourbench/pipeline/multi_hop_question_generation.py

```diff
@@ -0,0 +1,510 @@
+                        thought_process = qap.get("thought_process", "")
+                        if not isinstance(thought_process, str):
+                            thought_process = str(thought_process)
+
+                        cits = qap.get("citations", [])
+                        if not isinstance(cits, list):
+                            logger.warning("Citations for doc_id={} is not a list. Converting to empty list.", doc_id)
+                            cits = []
+
+                        # Ensure all citation items are strings
+                        cits = [str(c) for c in cits]
+
+                        row_obj = MultiHopQuestionRow(
+                            document_id=doc_id,
+                            source_chunk_ids=source_chunk_ids,
+                            question=question_text,
+                            self_answer=self_answer,
+                            estimated_difficulty=diff_val,
+                            self_assessed_question_type=qtype,
+                            generating_model=model_name,
+                            thought_process=thought_process,
+                            citations=cits,
+                            raw_response=raw_resp,
+                        )
+                        final_multi_hop_questions.append(row_obj.__dict__)
+
+                    except Exception as pair_error:
+                        logger.warning("Error processing QA pair for doc_id={}, skipping pair: {}", doc_id, pair_error)
+                        continue
+
+        if not final_multi_hop_questions:
+            logger.warning("No valid multi-hop question rows produced. Exiting stage.")
+            return
+
+        logger.info("Constructing multi-hop question dataset with {} rows...", len(final_multi_hop_questions))
+
+        # Convert to Hugging Face Dataset
+        try:
+            col_keys = list(final_multi_hop_questions[0].keys())
+            dataset_dict = {k: [row[k] for row in final_multi_hop_questions] for k in col_keys}
+            final_question_dataset = Dataset.from_dict(dataset_dict)
+        except Exception as ds_error:
+            logger.error("Failed to create dataset from multi-hop question rows: {}", ds_error)
+            return
+
+        # Save final dataset
+        logger.info("Saving multi-hop question dataset as '{}'.", output_dataset_name)
+        save_dataset(
+            dataset=final_question_dataset,
+            step_name="multi_hop_question_generation",
+            config=config,
+            output_dataset_name=output_dataset_name,
+            output_subset=output_subset,
+        )
+        logger.success("Multi-hop question generation completed successfully.")
+
+    except Exception as outer_exc:
+        logger.error("Error in multi_hop_question_generation run function: {}", str(outer_exc))
+        logger.warning("Multi-hop question generation stage encountered errors.")
+
+
+def _extract_tag_content(text: str, tag: str) -> str:
```

Comment:
should go in an "utils/extraction.py" file, as it can be used by other files - same for the following fn